libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -0,0 +1,343 @@
+const float OA_VERY_GOOD_MARGIN = 10000;        // very good margin to use during initialisation
+
+/// Constructor
+AP_OABendyRuler::AP_OABendyRuler()
+{
+    _singleton = this;
+}
+
+// run background task to find best path and update avoidance_results
+// returns true and updates origin_new and destination_new if a best path has been found
+bool AP_OABendyRuler::update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new)
+{
+    // exit immediately if neither circular nor polygon fence is enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr != nullptr) {
+        if ((fence_ptr->get_enabled_fences() & (AC_FENCE_TYPE_CIRCLE | AC_FENCE_TYPE_POLYGON)) == 0) {
+            return false;
+        }
+    }
+
+    // bendy ruler always sets origin to current_loc
+    origin_new = current_loc;
+
+    // calculate bearing and distance to final destination
+    const float bearing_to_dest = current_loc.get_bearing_to(destination) * 0.01f;
+    const float distance_to_dest = current_loc.get_distance(destination);
+
+    // retrieve margin from fence
+    const float fence_margin = MIN(fence_ptr->get_margin(), 100.0f);
+
+    // lookahead distance is adjusted dynamically based on avoidance results
+    _current_lookahead = constrain_float(_current_lookahead, _lookahead * 0.5f, _lookahead);
+
+    // calculate lookahead dist and time for step1.  distance can be slightly longer than
+    // the distance to the destination to allow room to dodge after reaching the destination
+    const float lookahead_step1_dist = MIN(_current_lookahead, distance_to_dest + MIN(fence_margin / 2.0f, OA_LOOKAHEAD_PAST_DEST));
+
+    // calculate lookahead dist for step2
+    const float lookahead_step2_dist = _current_lookahead * OA_LOOKAHEAD_STEP2_RATIO;
+
+    // get ground course
+    float ground_course_deg;
+    if (ground_speed_vec.is_zero()) {
+        // with zero ground speed use vehicle's heading
+        ground_course_deg = AP::ahrs().yaw_sensor * 0.01f;
+    } else {
+        ground_course_deg = ground_speed_vec.angle();
+    }
+
+    // search in 5 degree increments around the vehicle alternating left
+    // and right. For each direction check if vehicle would avoid all obstacles
+    float best_bearing = bearing_to_dest;
+    bool have_best_bearing = false;
+    float best_margin = OA_VERY_BAD_MARGIN;
+    float best_margin_bearing = best_bearing;
+
+    for (uint16_t i = 0; i < 360 / OA_BEARING_INC; i++) {
+        float bearing_delta = i * OA_BEARING_INC * 0.5f;
+        if (i & 1) {
+            // we alternate between trying to the left and right of the target
+            bearing_delta = -bearing_delta;
+        }
+        // bearing that we are probing
+        const float bearing_test = wrap_180(bearing_to_dest + bearing_delta);
+        const float bearing_change_deg = wrap_180(bearing_test - ground_course_deg);
```

Comment:
yes, it's certainly possible that it might be noisy.  It's got a "ground_speed_vec.is_zero()" check further up but the EKF is unlikely to report an exactly zero speed.  It's more likely to be low.. so we could add some minimum speed in there instead.. maybe 0.2m/s?

Reply:
I've replaced the is_zero with a OA_BENDYRULER_LOW_SPEED_SQUARED definition of 0.04f (ie. 0.2m/s)

Reply:
This has definitely improved behaviour by the way.  It would sometimes become indecisive, i was unsure of the cause but I think this was it.